crates/sui-graphql-rpc/src/metrics.rs

```diff
@@ -1,15 +1,209 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-use prometheus::{register_histogram_with_registry, Histogram, Registry};
-
-#[derive(Clone, Debug)]
-pub struct RequestMetrics {
-    pub(crate) input_nodes: Histogram,
-    pub(crate) output_nodes: Histogram,
-    pub(crate) query_depth: Histogram,
-    pub(crate) query_payload_size: Histogram,
-    pub(crate) _db_query_cost: Histogram,
+use std::sync::Arc;
+
+use async_graphql::{PathSegment, ServerError};
+use prometheus::{
+    register_histogram_vec_with_registry, register_histogram_with_registry,
+    register_int_counter_vec_with_registry, Histogram, HistogramVec, IntCounterVec, Registry,
+};
+use std::fmt::Write;
+
+use crate::error::code;
+
+#[derive(Clone)]
+pub(crate) struct Metrics {
+    pub db_metrics: Arc<DBMetrics>,
+    pub request_metrics: Arc<RequestMetrics>,
+}
+
+impl Metrics {
+    pub(crate) fn new(registry: &Registry) -> Self {
+        let db_metrics = DBMetrics::new(&registry);
+        let request_metrics = RequestMetrics::new(&registry);
+
+        Self {
+            db_metrics: Arc::new(db_metrics),
+            request_metrics: Arc::new(request_metrics),
+        }
+    }
+
+    /// Updates the DB related metrics (latency, error, success)
+    pub(crate) fn observe_db_data(&self, db_latency: f64, succeeded: bool) {
+        self.db_metrics
+            .db_fetch_latency
+            .with_label_values(&["latency", "db"])
+            .observe(db_latency);
+        if succeeded {
+            self.db_metrics
+                .db_num_fetches
+                .with_label_values(&["success"])
+                .inc();
+        } else {
+            self.db_metrics
+                .db_num_fetches
+                .with_label_values(&["error"])
+                .inc();
+        }
+    }
+
+    /// Use this function to increment the number of errors due to bad input
+    pub(crate) fn inc_num_bad_input_errors(&self, label: &str) {
+        self.request_metrics
+            .num_bad_input_errors_by_path
+            // TODO when measuring cost related errors, there's no path
+            // not 100% how to label such an error?
+            .with_label_values(&[label])
+            .inc();
+    }
+
+    /// Use this function to increment the number of errors per path, and labels
+    /// them as either internal server error, bad user input, or any other error.
+    /// The path is the same as the path that the user sees.
+    pub(crate) fn inc_errors(&self, errors: Vec<ServerError>) {
+        for err in errors {
+            if let Some(ext) = err.extensions {
+                if let Some(code) = ext.get("code") {
+                    if let async_graphql_value::Value::String(val) = code.clone().into_value() {
+                        if val == code::INTERNAL_SERVER_ERROR {
+                            self.request_metrics
+                                .num_internal_errors_by_path
```

Comment:
Why not have the one metric, like request_error, and then multiple labels, bad_input_error, internal_error etc

Reply:
Ah yes, that's where I ended up in the end.